includes/class-dispatcher.php

```diff
@@ -73,7 +79,161 @@ public static function process_outbox( $id ) {
 			$activity->set_object( $activity->get_object()->get_id() );
 		}
 
+		if ( ! self::should_send_to_followers( $activity, $actor_id, $outbox_item ) ) {
+			return;
+		}
+
 		self::send_activity_to_followers( $activity, $actor_id, $outbox_item );
+		self::async_batch_processor( $activity, $actor_id, $outbox_item );
+	}
+
+	/**
+	 * Asynchronously runs batch processing routines.
+	 *
+	 * @param Activity $activity    The ActivityPub Activity.
+	 * @param int      $actor_id    The actor ID.
+	 * @param \WP_Post $outbox_item The Outbox item.
+	 */
+	public static function async_batch_processor( $activity, $actor_id, $outbox_item ) {
+		// Bail if the existing lock is still valid.
+		if ( self::is_locked( $outbox_item->ID ) ) {
+			// phpcs:ignore PHPCompatibility.FunctionUse.ArgumentFunctionsReportCurrentValue.NeedsInspection
+			\wp_schedule_single_event( time() + MINUTE_IN_SECONDS, 'activitypub_outbox_process_batch', \func_get_args() );
+			return;
+		}
+
+		self::lock( $outbox_item->ID );
+
+		$args = array(
+			'post_type'      => Followers::POST_TYPE,
+			'posts_per_page' => self::$batch_size,
+			'fields'         => 'ids',
+
+			// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
+			'meta_query'     => array(
+				'relation' => 'AND',
+				array(
+					'key'     => '_activitypub_inbox',
+					'compare' => 'EXISTS',
+				),
+				array(
+					'key'     => '_activitypub_inbox',
+					'value'   => '',
+					'compare' => '!=',
+				),
+				array(
+					'key'     => '_activitypub_processed_outbox_' . $outbox_item->ID,
+					'compare' => 'NOT EXISTS',
+				),
+			),
+		);
+
+		if ( self::maybe_add_inboxes_of_blog_user( $activity, $actor_id ) ) {
+			$args['meta_query'][] = array(
+				'relation' => 'OR',
+				array(
+					'key'   => '_activitypub_actor_id',
+					'value' => Actors::BLOG_USER_ID,
+				),
+				array(
+					'key'   => '_activitypub_actor_id',
+					'value' => $actor_id,
+				),
+			);
+		} else {
+			$args['meta_query'][] = array(
+				'key'   => '_activitypub_actor_id',
+				'value' => $actor_id,
+			);
+		}
+
+		// Query for the next batch of followers.
+		$followers = get_posts( $args );
+		$json      = $activity->to_json();
+
+		foreach ( $followers as $follower_id ) {
+			$inbox  = get_post_meta( $follower_id, '_activitypub_inbox', true );
+			$result = safe_remote_post( $inbox, $json, $actor_id );
+
+			// @TODO Handle errors more elegantly.
+			if ( wp_remote_retrieve_response_code( $result ) >= 400 ) {
+				$attempt = get_post_meta( $follower_id, '_activitypub_outbox_attempt_' . $outbox_item->ID, true );
+				$attempt = $attempt ? $attempt + 1 : 1;
+
+				if ( $attempt <= 3 ) {
+					// Log attempt and move on.
+					update_post_meta( $follower_id, '_activitypub_outbox_attempt_' . $outbox_item->ID, $attempt );
```

Comment:
is that a good idea? in the worst case this would add tens of thousands of metas.

Reply:
But only until they're processed. We should remove them once that happened. Even if "finished processing" means we tried 3 times unsuccessfully.

Reply:
Can't that kill shared hosting dbs? 1000++ db inserts in a few milli seconds?

Reply:
It's batches of 50 currently and this only runs on failed posts. How did you get to thousands?

Reply:
oh, I may have commented on the wrong line... but some lines later, the code also stores the processed ones!?!

Reply:
Correct, [they get deleted](https://github.com/Automattic/wordpress-activitypub/pull/1262/files#diff-f53a4c6f8c1e61852fbbdd1132034377c601fdabc3bebb9dd2ac9c834c41a9ddR180) once an outbox item finished processing.